src/ripple/basics/base_uint.h

```diff
@@ -177,22 +178,82 @@ class base_uint
         memcpy(data_.data(), data, bytes);
     }
 
+    // Helper function to initialize a base_uint from a std::string_view.
+    constexpr decltype(data_)
+    initFromStringView(std::string_view sv)
+    {
+        // Local lambda that converts a single hex char to four bits.
+        // Return is a uint32_t to avoid integer conversions.
+        auto hexCharToUInt = [](char c) -> std::uint32_t {
+            std::uint32_t ret = 0xFFu;
+            if (c < '0' || c > 'f')
+                Throw<std::range_error>("invalid hex character");
+
+            if (c >= 'a')
+                ret = static_cast<std::uint32_t>(c - 'a' + 0xA);
+            else if (c >= 'A')
+                ret = static_cast<std::uint32_t>(c - 'A' + 0xA);
+            else if (c <= '9')
+                ret = static_cast<std::uint32_t>(c - '0');
+
+            if (ret > 0xFu)
+                Throw<std::range_error>("invalid hex character");
+
+            return ret;
+        };
+
+        decltype(data_) ret{};
+
+        if (sv == "0")
+        {
+            return ret;
+        }
+
+        if (sv.size() != size() * 2)
+            Throw<std::invalid_argument>("invalid length for hex string");
+
+        auto out = ret.data();
+
+        auto in = sv.begin();
+
+        while (in != sv.end())
+        {
+            std::uint32_t accum = {};
+            accum |= hexCharToUInt(*in++) << 4;
+            accum |= hexCharToUInt(*in++) << 0;
+            accum |= hexCharToUInt(*in++) << 12;
+            accum |= hexCharToUInt(*in++) << 8;
+            accum |= hexCharToUInt(*in++) << 20;
+            accum |= hexCharToUInt(*in++) << 16;
+            accum |= hexCharToUInt(*in++) << 28;
+            accum |= hexCharToUInt(*in++) << 24;
+            *out++ = accum;
+        }
+        return ret;
```

Comment:
Is it possible to factor out some common function between this and `parseHex`?

Reply:
Good question.

I think that instead of factoring out common functionality, I'd simply make `parseHex(std::string_view)` be `constexpr`. Doing this will require making [`charUnHex`](https://github.com/ripple/rippled/blob/develop/src/ripple/basics/impl/strHex.cpp#L25-L47) be `constexpr` but that seems simple and, honestly, desirable in other contexts.

Reply:
I thought that wasn't possible because `constexpr` implies `const`, but it looks like that changed in C++14. In that case, I would prefer `constexpr parseHex` too.

Reply:
I can make `parseHex()` `constexpr`.  In the process most of the code between `parseHex()` and the constructor from `string_view` will be the same.  A `constexpr` `parseHex()` is not likely to be especially useful since it mutates data members.  This means `parseHex()` can be used `constexpr` as long as that mutation stays inside of a single `constexpr` context.  Once that context exits then any `base_uint` produced by that context becomes immutable.

There are some downsides to the shared code approach.  An important distinction between `parseHex()` and the constructor is that `parseHex()` does not throw on errors; the constructor does.  This means the code must be structured a bit differently.

It's also important to note that `base_uint`'s `iterator` and `data()` interfaces can never be used in a `constexpr` context because they require `reinterpret_cast`.  That's why the `constexpr` methods in `base_uint` operate on `std::uint32_t`, which is the actual data type inside the `std::array`.

I've pushed a commit showing what the `parseHex()` changes look like.  If you like it I can keep the commit.  Otherwise it's easy to back out.

Reply:
I was thinking of a different approach to `constexpr parseHex`. The member `data_` is a `std::array`, so we don't really gain anything by constructing it in an initializer instead of default constructing it and assigning into it from the body of `base_uint`'s constructor. I was picturing both `parseHex` and the constructor would call a third function that takes both the `std::string_view` input parameter and a `std::array<...>&` output parameter, returns some error type (`bool`, `ParseResult`, `std::exception_ptr`, whatever), and assigns into the array. `parseHex` would just return the same value, and the constructor would branch on that value to throw the right exception. That "third" function could actually just be `parseHex`, though I wouldn't mind it being a separate function.

Reply:
@thejohnfreeman, I'm not very clear on the distinction between what you're describing and what I've coded up.  May I suggest you code up the alternative you have in mind and we'll take a look?

In terms of initializing `base_uint::data_` in the constructor, all variables must always be in a known state in a constexpr context.  So if we allow `data_` to default initialize, it will initially be filled with all zeros.  Then we would overwrite those initial values in the constructor body.

To me it seems a bit more to the point to simply initialize `data_` with the right value in the first place.  But your milage may vary.

Reply:
[This](https://github.com/thejohnfreeman/rippled/commit/73ebc5333981305586d331f3fccc7fa277e117b2) is what I was thinking. It can still be changed to give a better error message, and to distinguish between errors of "too few digits" and "non-digit found", but I think it captures the idea of sharing the main body without too much wrangling over return code vs exception.

Most of the time, I too insist on initializing members in their initializer, but in the case of a `constexpr` constructor, I'm fine relaxing that rule to get code that I think is easier to read. The compiler [produces](https://godbolt.org/z/978T4n) the same thing in the end.

Reply:
@thejohnfreeman, thanks for the example.  Certainly both approaches do the job.

Personally, I like having the constructor distinguish between bad lengths and invalid characters in the error.  If we sacrifice that usability improvement then what you've coded up works well.  We could add the length/character distinction into the return value of `parseHex()`, but then we'd need to edit all of the callers of `parseHex()`.  I don't think that's worth it.

I'm also not keen on moving `hexCharToUint(char c, T& accum, T shift = 0)` out to be a library function and making it a template.  As a general function it should be receiving more extensive testing.  It also needs some guards on the various kinds of template parameters it can take.  I just don't suspect it will get much use out there.  I like to keep simple specialized functions, like the original version, in a local scope.  But that's just my personal preference.

But if you find your implementation is truly preferable then, cool.  Add appropriate template parameter guards to `hexCharToUint()` and add some unit tests and we can call it good.  I'm open to losing the distinction in error types.